### ISSUE TITLE
CommutationAnalysis is not idempotent

### ISSUE THREAD

**Issue by 1ucian0** [Qiskit/qiskit-terra#4998]
As @kdk explains in [this comment](https://github.com/Qiskit/qiskit-terra/issues/4984#issuecomment-683076229), the `CommutationAnalysis` pass is storing internal state and that (probably) creates invalid results when `run` is called more than once.

**Comment by 1ucian0** [comment]
In a deeper look it seems safe to share `CommutationAnalysis.cache`. I will explore a bit longer.

`CommutationAnalysis.gates_on_wire` is not used and can be removed.

**Comment by kdk** [comment]
> In a deeper look it seems safe to share `CommutationAnalysis.cache`. I will explore a bit longer.
> 
> `CommutationAnalysis.gates_on_wire` is not used and can be removed (PR #5000)

Agree, since the cache only records commutation relations which are globally true (and not specific to a particular DAG or call to `run`), this is okay.
